rotate.go

```diff
@@ -602,6 +614,47 @@ func (r *Rotator) rollback(ctx context.Context, creds db.NewPassword, rotationSt
 	return errRotationFailed // always return this error
 }
 
+// checks that secret have been replicated to all replica regions
+// this is necessary between multiple calls of UpdateSecretVersionStage
+// to guard against arace condition in AWS that leaves secret replication
+// stuck indefinitely.
+func (r *Rotator) checkSecretReplicationStatus() error {
+	log.Println("checking secret replication status")
+	waitDuration := DEFAULT_REPLICATION_WAIT_DURATION
+	if r.replicationWaitDuration > 0 {
+		waitDuration = r.replicationWaitDuration
+	}
+
+	startTime := time.Now()
+	for time.Now().Sub(startTime) < waitDuration {
+		secret, err := r.sm.DescribeSecret(&secretsmanager.DescribeSecretInput{
+			SecretId: aws.String(r.secretId),
+		})
+		if err != nil {
+			return err
+		}
+		if secret == nil {
+			return fmt.Errorf("expected an non null secret for secretId %v but received null", r.secretId)
+		}
+		replicationSyncComplete := true
+		for _, status := range secret.ReplicationStatus {
+			if status == nil {
+				continue
```

Comment:
Don't continue here; you need to goto the sleep or something. Else, this might enter an uncontrolled loop and burn massive Lambda CPU time/credits.

Reply:
Although Go usually "reads down the left", this is one case where the error stuff might read down the left (instead of the normal path):
```go
for ... {
    if status != nil && *status.Status == secretsmanager.StatusTypeInSync {
        log.Print...
        return nil // success
    }
    // All the error handling stuff
    time.Sleep()
}
```

Whenever I do this (make normal path _not_ read down the left, I call out the exception with a comment like `// success` so people familiar with Go know we're doing an early return instead of reading down the left.

Reply:
Updated